execution/block-partitioner/src/sharded_block_partitioner/mod.rs

```diff
@@ -0,0 +1,685 @@
+            num_shards
+        );
+        assert!(num_shards > 0, "num_partitioning_shards must be > 0");
+        // create channels for cross shard messages across all shards. This is a full mesh connection.
+        // Each shard has a vector of channels for sending messages to other shards and
+        // a vector of channels for receiving messages from other shards.
+        let mut messages_txs = vec![];
+        let mut messages_rxs = vec![];
+        for _ in 0..num_shards {
+            messages_txs.push(vec![]);
+            messages_rxs.push(vec![]);
+            for _ in 0..num_shards {
+                let (messages_tx, messages_rx) = std::sync::mpsc::channel();
+                messages_txs.last_mut().unwrap().push(messages_tx);
+                messages_rxs.last_mut().unwrap().push(messages_rx);
+            }
+        }
+        let mut control_txs = vec![];
+        let mut result_rxs = vec![];
+        let mut shard_join_handles = vec![];
+        for (i, message_rxs) in messages_rxs.into_iter().enumerate() {
+            let (control_tx, control_rx) = std::sync::mpsc::channel();
+            let (result_tx, result_rx) = std::sync::mpsc::channel();
+            control_txs.push(control_tx);
+            result_rxs.push(result_rx);
+            shard_join_handles.push(spawn_partitioning_shard(
+                i,
+                control_rx,
+                result_tx,
+                message_rxs,
+                messages_txs.iter().map(|txs| txs[i].clone()).collect(),
+            ));
+        }
+        Self {
+            num_shards,
+            control_txs,
+            result_rxs,
+            shard_threads: shard_join_handles,
+        }
+    }
+
+    // reorders the transactions so that transactions from the same sender always go to the same shard.
+    // This places transactions from the same sender next to each other, which is not optimal for parallelism.
+    // TODO(skedia): Improve this logic to shuffle senders
+    fn reorder_txns_by_senders(
+        &self,
+        txns: Vec<AnalyzedTransaction>,
+    ) -> Vec<Vec<AnalyzedTransaction>> {
+        let approx_txns_per_shard = (txns.len() as f64 / self.num_shards as f64).ceil() as usize;
+        let mut sender_to_txns = HashMap::new();
+        let mut sender_order = Vec::new(); // Track sender ordering
+
+        for txn in txns {
+            let sender = txn.sender().unwrap();
+            if !sender_to_txns.contains_key(&sender) {
+                sender_order.push(sender); // Add sender to the order vector
+            }
+            sender_to_txns
+                .entry(sender)
+                .or_insert_with(Vec::new)
+                .push(txn);
+        }
+
+        let mut result = Vec::new();
+        result.push(Vec::new());
+
+        for sender in sender_order {
+            let txns = sender_to_txns.remove(&sender).unwrap();
+            let txns_in_shard = result.last().unwrap().len();
+
+            if txns_in_shard < approx_txns_per_shard {
```

Comment:
The distribution can be slightly more balanced with `if txns_in_shard + txns.len() < approx_txns_per_shard {`?

Reply:
Can you elaborate why the above will lead to a better balanced distribution?

Reply:
Actually not much difference... please ignore this comment :D